File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Public/Events/EventBus.swift
//
//  EventBus.swift
//  RunAnywhere SDK
//
//  Simple pub/sub for SDK events.
//

import Combine
import Foundation

// MARK: - Event Bus

/// Central event bus for SDK-wide event distribution.
///
/// Subscribe to events by category or to all events:
/// ```swift
/// // Subscribe to all events
/// EventBus.shared.events
///     .sink { event in print(event.type) }
///
/// // Subscribe to specific category
/// EventBus.shared.events(for: .llm)
///     .sink { event in print(event.type) }
/// ```
public final class EventBus: @unchecked Sendable {

    // MARK: - Singleton

    public static let shared = EventBus()

    // MARK: - Publishers

    private let subject = PassthroughSubject<any SDKEvent, Never>()

    /// All events publisher
    public var events: AnyPublisher<any SDKEvent, Never> {
        subject.eraseToAnyPublisher()
    }

    // MARK: - Initialization

    private init() {}

    // MARK: - Publishing

    /// Publish an event to all subscribers
    public func publish(_ event: any SDKEvent) {
        subject.send(event)
    }

    // MARK: - Filtered Subscriptions

    /// Get events for a specific category
    public func events(for category: EventCategory) -> AnyPublisher<any SDKEvent, Never> {
        subject
            .filter { $0.category == category }
            .eraseToAnyPublisher()
    }

    /// Subscribe to events with a closure
    public func on(_ handler: @escaping (any SDKEvent) -> Void) -> AnyCancellable {
        subject.sink { event in
            handler(event)
        }
    }

    /// Subscribe to events of a specific category
    public func on(
        _ category: EventCategory,
        handler: @escaping (any SDKEvent) -> Void
    ) -> AnyCancellable {
        events(for: category).sink { event in
            handler(event)
        }
    }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/commonMain/kotlin/com/runanywhere/sdk/public/events/EventBus.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Simple pub/sub for SDK events.
 *
 * Mirrors Swift EventBus.swift exactly.
 */

package com.runanywhere.sdk.public.events

import com.runanywhere.sdk.foundation.SDKLogger
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.filter

/**
 * Central event bus for SDK-wide event distribution.
 *
 * Subscribe to events by category or to all events:
 *
 * ```kotlin
 * // Subscribe to all events
 * EventBus.events.collect { event ->
 *     println(event.type)
 * }
 *
 * // Subscribe to specific category
 * EventBus.events(EventCategory.LLM).collect { event ->
 *     println(event.type)
 * }
 * ```
 *
 * Mirrors Swift EventBus exactly.
 */
object EventBus {
    // MARK: - Publishers

    private val logger = SDKLogger.shared

    private val _events =
        MutableSharedFlow<SDKEvent>(
            replay = 0,
            extraBufferCapacity = 64,
        )

    /** All events flow */
    val events: Flow<SDKEvent> = _events.asSharedFlow()

    // MARK: - Publishing

    /**
     * Publish an event to all subscribers.
     */
    fun publish(event: SDKEvent) {
        logger.debug("Publishing event: ${event.type} (category: ${event.category.value})")
        _events.tryEmit(event)
    }

    // MARK: - Filtered Subscriptions

    /**
     * Get events for a specific category.
     */
    fun events(category: EventCategory): Flow<SDKEvent> {
        return events.filter { it.category == category }
    }

    /**
     * Get events of a specific type.
     */
    inline fun <reified T : SDKEvent> eventsOfType(): Flow<T> {
        return events.filter { it is T } as Flow<T>
    }

    // MARK: - Convenience Methods

    /**
     * Get LLM events.
     */
    val llmEvents: Flow<SDKEvent>
        get() = events(EventCategory.LLM)

    /**
     * Get STT events.
     */
    val sttEvents: Flow<SDKEvent>
        get() = events(EventCategory.STT)

    /**
     * Get TTS events.
     */
    val ttsEvents: Flow<SDKEvent>
        get() = events(EventCategory.TTS)

    /**
     * Get model events.
     */
    val modelEvents: Flow<SDKEvent>
        get() = events(EventCategory.MODEL)

    /**
     * Get error events.
     */
    val errorEvents: Flow<SDKEvent>
        get() = events(EventCategory.ERROR)

    /**
     * Get SDK lifecycle events.
     */
    val sdkEvents: Flow<SDKEvent>
        get() = events(EventCategory.SDK)

    /**
     * Get RAG events.
     */
    val ragEvents: Flow<SDKEvent>
        get() = events(EventCategory.RAG)
}
